### ISSUE TITLE
Unknown error calling sqlite3_step (10: disk I/O error) rs

### ISSUE THREAD

**Issue by tedrog36** [ccgus/fmdb#535]
Just started getting error or at least getting it more often on iOS10.  

See this link for more info: https://github.com/marcoarment/FCModel/issues/105

Any ideas of what might be happening?  The code is currently built with FMDB 2.5.

Thanks.

**Comment by ccgus** [comment]
Show all the backtraces please.

**Comment by tedrog36** [comment]
In this stacktrace thread 23 is sitting on this statement in nextWithError:
        // wtf?
        NSLog(@"Unknown error calling sqlite3_step (%d: %s) rs", rc, sqlite3_errmsg([_parentDB sqliteHandle]));

[StackTrace.txt](https://github.com/ccgus/fmdb/files/481002/StackTrace.txt)

**Comment by ccgus** [comment]
I haven't looked at FCModel, so I can only guess as to what's going on, but it looks like both thread 22 & 23 are doing database operations at the same time, which is a bad idea.

**Comment by tedrog36** [comment]
That is what I thought at first but thread 22 and 23 use the same operation queue and it appears that 22 is waiting for 23 to finish.

**Comment by tedrog36** [comment]
Gus - I am just learning this codebase I am debugging.  I realize that it is writing images to the database as JPEG NSData.

Any experience with issues around doing this? 

My current hypothesis is that writing these large images to the database is driving this problem.

**Comment by ccgus** [comment]
Acorn uses FMDB, and writes some very large images to databases (100mp even), so I don't think that's it. I think there's a threading error somewhere.

**Comment by tedrog36** [comment]
I update to the latest FCModel which puts all database requests on the main UI thread so I don't think it is that.

**Comment by ccgus** [comment]
You still get the problem when all requests are on the main thread? Can you show all the backtraces when that happens?

**Comment by tedrog36** [comment]
See attached stacktrace from Crashlytics.
[crashlytics_stacktrace.txt](https://github.com/ccgus/fmdb/files/485668/crashlytics_stacktrace.txt)

**Comment by ccgus** [comment]
This isn't the same error as before. The app is dying because of a thrown exception.

**Comment by tedrog36** [comment]
Yes.  "Query failed with SQLite error 10: disk I/O error"

Which is the same root cause as in the previous stack trace.

**Comment by ccgus** [comment]
I don't know what's causing, sorry. A reproducible case (in source code form) would help.

**Comment by tedrog36** [comment]
Yeah, I know.  Unfortunately, I have not found a way to reproduce it.

Thanks anyway.  I appreciate the time you took to check out the stack traces.

Ted

**Comment by robertmryan** [comment]
Error 10, `SQLITE_IOERR`, is a result of "Some kind of disk I/O error occurred". As [the docs say](http://sqlite.org/rescode.html#ioerr):

> The `SQLITE_IOERR` result code says that the operation could not finish because the operating system reported an I/O error.

> A full disk drive will normally give an `SQLITE_FULL` error rather than an `SQLITE_IOERR` error.

> There are many different extended result codes for I/O errors that identify the specific I/O operation that failed.

It sure sounds like some basic file issue. I'd suggest checking out `sqlite3_extended_errcode`, and see if that's illuminating. @ccgus - it strikes me that we should expose that API.

I assume your database is definitely in a writable directory (i.e. not the bundle, but rather temp, documents or caches folder)?

But in answer to your question, there are no known issues in writing data objects to a database (though it's much slower than saving the images in the file system and only saving a path in the database).

**Comment by tedrog36** [comment]
I removed the saving of the images to the database but still got the error.  
I was finally able to reproduce this is the debugger.  I have attached the stack trace, the query, and the values.  I don't understand how the primary key is handled but its value of "2327685398450371228" looked suspicious to me.
Any ideas?

[stacktrace_debug.txt](https://github.com/ccgus/fmdb/files/487693/stacktrace_debug.txt)

**Comment by yogesh-zendrive** [comment]
Any update on this issue? I am getting a lot of logs like these in our continuous unit-test environment. This behaves kind of like a locked database.

**Comment by tedrog36** [comment]
@yogesh-zendrive I would like to say I found a fix for this but I ended up using a completely different mechanism for saving the data.

**Comment by groue** [comment]
One possible reason for SQLITE_IOERR 10 "disk I/O error" to occur is when data protection prevents SQLite from accessing the database file on a locked device.

Experience shows that data protection can also raise SQLITE_AUTH 23 "not authorized".

Quoting [GRDB FAQ](https://github.com/groue/GRDB.swift/blob/master/README.md#sqlite-error-10-disk-io-error-sqlite-error-23-not-authorized):

> Those errors may be the sign that SQLite can't access the database due to [data protection](https://github.com/groue/GRDB.swift/blob/master/README.md#data-protection).
>
> When your application should be able to run in the background on a locked device, it has to catch this error, and, for example, wait for [UIApplicationDelegate.applicationProtectedDataDidBecomeAvailable(_:)](https://developer.apple.com/reference/uikit/uiapplicationdelegate/1623044-applicationprotecteddatadidbecom) or [UIApplicationProtectedDataDidBecomeAvailable](https://developer.apple.com/reference/uikit/uiapplicationprotecteddatadidbecomeavailable) notification and retry the failed database operation.
> 
> This error can also be prevented altogether by using a more relaxed [file protection](https://developer.apple.com/reference/foundation/filemanager/1653059-file_protection_values).

**Comment by yogesh-zendrive** [comment]
@groue Thanks for the info. This is happening in continuous testing environment that uses simulator to run tests, is the problem you mentioned possible in the simulator?
@tedrog36 Did you figure out root of the problem? What mechanism are you using now to save the data?

NOTE: This problem is becoming bigger and bigger as I am adding more extensive tests for algorithm performance. Today, the tests failed because of timeout and the log file contained more than 8 million lines of "Unknown error calling sqlite3_step (10: disk I/O error) rs" log.

**Comment by tedrog36** [comment]
@yogesh-zendrive I did not figure out the root of the problem after many hours of chasing it.  The app I was working on needed to save very little data so I just used file serialization and deserialized the data into an NSDictionary for quick lookup.

**Comment by groue** [comment]
@yogesh-zendrive I'm not sure, but Google will certainly tell you.

**Comment by yogesh-zendrive** [comment]
I figured out the root cause for my case. I had two instances of FMDatabaseQueue open at the same time. There was an issue in my unit-test teardown method, the old FMDatabaseQueue object was not getting deallocated or closed immediately and a new object was allocated in parallel.
This was leading the database to lock and the test was getting stuck in "Unknown error calling sqlite3_step (10: disk I/O error) rs" error loop.
@tedrog36 @groue Thanks for the help.

**Comment by valdemiro-vieira** [comment]
@yogesh-zendrive Can you elaborate on the issue in your unit tests? I am facing the same error that happens only in an integration test and I did not find out the problem.

**Comment by algrid** [comment]
I had a similar issue because I was deleting my database file in tests, but kept using the `FMDatabaseQueue` object that I previously created.
